core/src/checks/rpc_checks.c

```diff
@@ -0,0 +1,202 @@
+#include "checks.h"
+#include "config.h"
+#include "log.h"
+#include "nanopb_stream.h"
+#include "print.h"
+#include "rand.h"
+#include "rpc.h"
+
+#include <assert.h>
+#include <pb_decode.h>
+#include <pb_encode.h>
+#include <squareup/subzero/internal.pb.h>
+
+static size_t get_serialized_message_size(const InternalCommandRequest* const cmd) {
+  pb_ostream_t stream = PB_OSTREAM_SIZING;
+  if (!pb_encode_delimited(&stream, InternalCommandRequest_fields, cmd)) {
+    ERROR("%s: pb_encode_delimited() failed: %s", __func__, PB_GET_ERROR(&stream));
+    return 0;
+  }
+  return stream.bytes_written;
+}
+
+int verify_rpc_oversized_message_rejected(void) {
+  int result = 0;
+  uint8_t* serialized_request = NULL;
+  uint8_t* serialized_response = NULL;
+
+  InternalCommandRequest cmd = InternalCommandRequest_init_default;
+  cmd.version = VERSION;
+  cmd.wallet_id = 1; // dummy value
+  cmd.which_command = InternalCommandRequest_InitWallet_tag;
+  static_assert(
+      sizeof(cmd.command.InitWallet.random_bytes.bytes) == MASTER_SEED_SIZE,
+      "MASTER_SEED_SIZE must equal sizeof(cmd.command.InitWallet.random_bytes.bytes)");
+  cmd.command.InitWallet.random_bytes.size = MASTER_SEED_SIZE;
+  random_buffer(cmd.command.InitWallet.random_bytes.bytes, MASTER_SEED_SIZE);
+
+  size_t serialized_size = get_serialized_message_size(&cmd);
+  if (serialized_size == 0) {
+    result = -1;
+    goto out;
+  }
+
+  // Note that we allocate 1 extra byte because we'll be extending the message.
+  serialized_request = (uint8_t*) calloc(1, serialized_size + 1);
+  if (NULL == serialized_request) {
+    ERROR("%s: calloc(1, %zu) failed", __func__, serialized_size + 1);
+    result = -1;
+    goto out;
+  }
+
+  pb_ostream_t ostream = pb_ostream_from_buffer(serialized_request, serialized_size);
+  if (!pb_encode_delimited(&ostream, InternalCommandRequest_fields, &cmd)) {
+    ERROR("%s: pb_encode_delimited() failed: %s", __func__, PB_GET_ERROR(&ostream));
+    result = -1;
+    goto out;
+  }
+
+  // Helper macro used to check our assumptions in the gnarly protobuf mangling code below
+#define ASSERT_BYTE_EQUALS(buf, idx, expected_val)                        \
+  do {                                                                    \
+    const uint8_t* buf_ = (buf);                                          \
```

Comment:
is the trailing _ a c convention?

Reply:
preprocessor as in using MACROS @oreparaz?

Reply:
Removing the macro in the latest version, use a helper function instead. Code ends up being more verbose though.